truffle/com.oracle.truffle.api.object.dsl.test/src/com/oracle/truffle/api/object/dsl/test/InheritanceTest.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2016, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.object.dsl.test;
+
+import org.junit.Test;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.ObjectType;
+import com.oracle.truffle.api.object.dsl.Layout;
+
+import java.math.BigInteger;
+
+import org.junit.Assert;
+
+public class InheritanceTest {
+
+    @Layout
+    public interface BaseLayout {
+
+        DynamicObject createBase(int a);
+
+        boolean isBase(DynamicObject object);
+
+        boolean isBase(Object object);
+
+        boolean isBase(ObjectType objectType);
+
+        int getA(DynamicObject object);
+
+        void setA(DynamicObject object, int value);
+
+    }
+
+    @Layout
+    public interface MiddleLayout extends BaseLayout {
+
+        DynamicObject createMiddle(int a, String b);
+
+        boolean isMiddle(DynamicObject object);
+
+        boolean isMiddle(Object object);
+
+        boolean isMiddle(ObjectType objectType);
+
+        String getB(DynamicObject object);
+
+        void setB(DynamicObject object, String value);
+
+    }
+
+    @Layout
+    public interface TopLayout extends MiddleLayout {
+
+        DynamicObject createTop(int a, String b, BigInteger c);
+
+        boolean isTop(DynamicObject object);
+
+        boolean isTop(Object object);
+
+        boolean isTop(ObjectType objectType);
+
+        BigInteger getC(DynamicObject object);
+
+        void setC(DynamicObject object, BigInteger value);
+
+    }
+
+    private static final BaseLayout BASE_LAYOUT = BaseLayoutImpl.INSTANCE;
+    private static final MiddleLayout MIDDLE_LAYOUT = MiddleLayoutImpl.INSTANCE;
+    private static final TopLayout TOP_LAYOUT = TopLayoutImpl.INSTANCE;
+
```

Comment:
Yes, I know. One always needs to process the whole package. I thought I even wrote a blog post about that, but I cannot find it. I used the code at http://hg.netbeans.org/html4j/file/ed4b25eb66f3/boot/src/main/java/org/netbeans/html/boot/impl/JavaScriptProcesor.java which also needs to generate single `$JsCallbacks$` file per package.